File: 994/main.go
package main

import (
	"fmt"
)

type Grid [][]int

func CloneEmpty(g Grid) Grid {
	m := len(g)
	n := len(g[0])
	newGrid := make([][]int, m)
	for i := 0; i < m; i++ {
		newGrid[i] = make([]int, n)
	}
	return newGrid
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func process(grid Grid, i, j, m, n, d int, ng Grid) {
	v := grid[i][j]
	grid[i][j] = 2
	if ng[i][j] > 0 {
		ng[i][j] = min(ng[i][j], d)
	} else {
		ng[i][j] = d
	}
	if i > 0 && grid[i-1][j] == 1 {
		process(grid, i-1, j, m, n, d+1, ng)
	}
	if i < m-1 && grid[i+1][j] == 1 {
		process(grid, i+1, j, m, n, d+1, ng)
	}
	if j > 0 && grid[i][j-1] == 1 {
		process(grid, i, j-1, m, n, d+1, ng)
	}
	if j < n-1 && grid[i][j+1] == 1 {
		process(grid, i, j+1, m, n, d+1, ng)
	}
	grid[i][j] = v
}

func orangesRotting(grid [][]int) int {
	if len(grid) < 1 {
		return 0
	}
	m, n := len(grid), len(grid[0])
	ng := CloneEmpty(grid)
	for i := 0; i < m; i++ {
		for j := 0; j < n; j++ {
			if grid[i][j] == 2 {
				process(grid, i, j, m, n, 0, ng)
			}
		}
	}
	maxCount := 0
	for i := 0; i < m; i++ {
		for j := 0; j < n; j++ {
			if grid[i][j] == 1 && ng[i][j] < 1 {
				return -1
			} else {
				if ng[i][j] > maxCount {
					maxCount = ng[i][j]
				}
			}
		}
	}
	return maxCount
}

func main() {
	input := [][]int{{2, 1, 1}, {1, 1, 0}, {0, 1, 1}}
	fmt.Println(orangesRotting(input))
	input = [][]int{{2, 1, 1}, {0, 1, 1}, {1, 0, 1}}
	fmt.Println(orangesRotting(input))
	input = [][]int{{2, 1, 1}, {1, 1, 0}, {0, 1, 2}}
	fmt.Println(orangesRotting(input))
	input = [][]int{{0, 2}}
	fmt.Println(orangesRotting(input))
	input = [][]int{{}}
	fmt.Println(orangesRotting(input))
	input = [][]int{{0}}
	fmt.Println(orangesRotting(input))
	input = [][]int{{2}}
	fmt.Println(orangesRotting(input))
	input = [][]int{{1, 2}}
	fmt.Println(orangesRotting(input))

	input = [][]int{{1}, {2}, {1}, {1}}
	fmt.Println(orangesRotting(input))

	input = [][]int{{0, 0, 1, 2}, {2, 0, 1, 1}}
	fmt.Println(orangesRotting(input))

}

File: 994/src/main.rs
// https://leetcode.com/problems/rotting-oranges/

pub struct Solution();

type Grid = Vec<Vec<i32>>;

impl Solution {
    fn has_neighbor_with(grid: &[Vec<i32>], i: usize, j: usize) -> bool {
        let m = grid.len();
        let n = grid[0].len();
        if j > 0 && grid[i][j - 1] == 2 {
            return true;
        }
        if j < n - 1 && grid[i][j + 1] == 2 {
            return true;
        }
        if i > 0 && grid[i - 1][j] == 2 {
            return true;
        }
        return i < m - 1 && grid[i + 1][j] == 2;
    }
    fn process_minute(grid: &[Vec<i32>]) -> Option<Grid> {
        let mut new_grid: Grid = grid.iter().cloned().collect();
        let m = new_grid.len();
        let n = new_grid[0].len();
        let mut changed = false;
        for i in 0..m {
            for j in 0..n {
                if new_grid[i][j] == 1 && Solution::has_neighbor_with(grid, i, j) {
                    changed = true;
                    new_grid[i][j] = 2
                }
            }
        }
        if changed {
            return Some(new_grid);
        }
        return None;
    }

    fn check_impossible(grid: &[Vec<i32>]) -> bool {
        let m = grid.len();
        let n = grid[0].len();
        for i in 0..m {
            for j in 0..n {
                if grid[i][j] == 1 {
                    return true;
                }
            }
        }
        false
    }

    pub fn oranges_rotting(grid: Grid) -> i32 {
        let mut c = 0;
        let mut g = grid;
        while let Some(grid) = Solution::process_minute(&g) {
            g = grid;
            c += 1
        }
        if Self::check_impossible(&g) {
            return -1;
        }
        c
    }
}

fn main() {
    let input: Grid = vec![vec![2, 1, 1], vec![1, 1, 0], vec![0, 1, 1]];
    // eprintln!("output = {:#?}", Solution::oranges_rotting(input));
    assert_eq!(Solution::oranges_rotting(input), 4);

    let input: Grid = vec![vec![2, 1, 1], vec![0, 1, 1], vec![1, 0, 1]];
    debug_assert_eq!(Solution::oranges_rotting(input), -1);

    let input: Grid = vec![vec![0, 2]];
    assert_eq!(Solution::oranges_rotting(input), 0);

    let input: Grid = vec![vec![0]];
    assert_eq!(Solution::oranges_rotting(input), 0);

    let input: Grid = vec![vec![0, 1], vec![2, 0]];
    assert_eq!(Solution::oranges_rotting(input), -1);

    let input: Grid = vec![vec![2], vec![2], vec![1], vec![0], vec![1], vec![1]];
    debug_assert_eq!(Solution::oranges_rotting(input), -1);

    let input: Vec<Vec<i32>> = vec![vec![1, 2]];
    debug_assert_eq!(Solution::oranges_rotting(input), 1);

    let input: Vec<Vec<i32>> = vec![vec![2, 1, 0, 2]];
    debug_assert_eq!(Solution::oranges_rotting(input), 1);
}
